### ISSUE TITLE
Event Loop?

### ISSUE THREAD

**Issue by nick-thompson** [nick-thompson/blueprint#20]
Right now there's no real notion of an event loop. Timers are implemented by a simple map from id (an auto-incremented int) to function with timestamp and the JUCE implementation just runs an interrupt every 4ms to execute any ready timers. This "works" but it's not great, and it won't scale nicely to something like a `fetch` polyfill which needs to push work off to the cpp side for actually collecting the response and calling back into js with the result.

**Comment by nick-thompson** [comment]
Thinking about a real event loop implementation in JUCE; essentially we just need a few things: a way to hold references to js callback functions, a timers handler, and some kind of thread pool for synchronous tasks.

* Callback references: https://wiki.duktape.org/howtonativepersistentreferences
    * So basically push a key/value pair into one of the duktape stash where key is some id (auto incremented integer or uuid maybe), and value is the callback. Can use this same id for `clearTimeout()` and related functionality
* Timers: `juce::Timer::callAfterDelay` with some wrapper code around registering/retrieving the callback
    * Wondering if it might be more effective to have a TimeSliceThread/TimeSliceClient that runs through the timers so that we don't drown the message queue on the message thread with timers
* Work queue: https://docs.juce.com/master/classThreadPool.html
    * This one seems pretty straightforward; register a callback with the stash, kick off some lambda into the threadpool and when we get the result call the callback

**Comment by nick-thompson** [comment]
Thinking about this some more, especially now that #9 is nearly complete. The new `EcmascriptEngine` class should now be the place for any sort of event loop implementation. I don't know that we necessarily need a beast of an event loop, I think it might be sufficient to install a couple native methods in EcmascriptEngine for `setTimeout`/`setInterval`/`clearTimeout`/`clearInterval` which just store a pointer to the callback in the duktape global stash, assigns an id, and pushes some struct into a timers queue. Then run a `juce::ThreadPool` where one job is a `TimersJob` that checks the expiry on each timer in the queue and prompts the `EcmascriptEngine` to call the callbacks and clear the structs when a timer is up. That `TimersJob` can mark itself as always needing to run again so we have a background thread just constantly keeping an eye on the timers. Then we can leverage the `ThreadPool`'s remaining capacity for things like a `fetch` polyfill

**Comment by nick-thompson** [comment]
See https://github.com/whoshuu/cpr for the fetch polyfill. Seems better equipped and easier to use than juce's URL class

**Comment by nick-thompson** [comment]
After #188 I think it's fair to call this done. Timers are the priority, fetch is a nice-to-have that, as of yet, doesn't seem to be a big want. I'm going to close this now, and we'll open a dedicated task to track a `fetch` implementation when it becomes more relevant
